File: jwt/go/jwt_one_route.go
package main

import (
	"fmt"
	"net/http"
	"time"

	"github.com/gorilla/mux"
	"github.com/golang-jwt/jwt/v5"
)

var jwtKey = []byte("your_secret_key")

type Claims struct {
	Username string `json:"username"`
	jwt.RegisteredClaims
}

func GenerateToken(username string) (string, error) {
	expirationTime := time.Now().Add(24 * time.Hour)
	claims := &Claims{
		Username: username,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(expirationTime),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(jwtKey)
}

func ValidateToken(tknStr string) (*Claims, error) {
	claims := &Claims{}
	tkn, err := jwt.ParseWithClaims(tknStr, claims, func(token *jwt.Token) (interface{}, error) {
		return jwtKey, nil
	})

	if err != nil {
		return nil, err
	}
	if !tkn.Valid {
		return nil, jwt.ErrSignatureInvalid
	}
	return claims, nil
}

func handleLogin(w http.ResponseWriter, r *http.Request) {
	username := r.PostFormValue("username")
	password := r.PostFormValue("password")

	// For demonstration, we're assuming "admin"/"password" as valid credentials.
	//Simulating user authentication
	username_database := "admin"
	password_database := "password"
	fmt.Println("Username:", username)
    fmt.Println("Password:", password)
	if username != username_database || password != password_database {
		http.Error(w, "Invalid credentials", http.StatusUnauthorized)
		return
	}

	token, err := GenerateToken(username)
	if err != nil {
		http.Error(w, "Failed to generate token", http.StatusInternalServerError)
		return
	}

	http.SetCookie(w, &http.Cookie{
		Name:    "token",
		Value:   token,
		Expires: time.Now().Add(24 * time.Hour),
	})
}

func handleProtectedRoute(w http.ResponseWriter, r *http.Request) {
	c, err := r.Cookie("token")
	if err != nil {
		if err == http.ErrNoCookie {
			http.Error(w, "No token", http.StatusUnauthorized)
			return
		}
		http.Error(w, "Bad Request", http.StatusBadRequest)
		return
	}

	tokenStr := c.Value
	claims, err := ValidateToken(tokenStr)
	if err != nil {
		http.Error(w, "Invalid token", http.StatusUnauthorized)
		return
	}

	w.Write([]byte(fmt.Sprintf("Hello, %s!", claims.Username)))
}

func main() {
	r := mux.NewRouter()

	r.HandleFunc("/login", handleLogin).Methods("POST")
	r.HandleFunc("/protected", handleProtectedRoute).Methods("GET")
	fmt.Println("JWT Server One Route started at :8080")
	http.ListenAndServe(":8080", r)
	
}

File: jwt/node/jwt_one_route.js
const express = require('express');
const jwt = require('jsonwebtoken');
const multer = require('multer');
const cookieParser = require('cookie-parser');

const app = express();
const jwtKey = 'your_secret_key';
const jwtExpirySeconds = 86400; // 24 hours

const upload = multer();
app.use(cookieParser());

function generateToken(username) {
    const token = jwt.sign({ username }, jwtKey, {
        algorithm: 'HS256',
        expiresIn: jwtExpirySeconds,
    });
    return token;
}

function validateToken(token) {
    try {
        const payload = jwt.verify(token, jwtKey);
        return payload;
    } catch (e) {
        return null;
    }
}

app.post('/login', upload.none(), (req, res) => {
    const { username, password } = req.body;

    // For demonstration, we're assuming "admin"/"password" as valid credentials.
    const username_database = 'admin';
    const password_database = 'password';

    if (username !== username_database || password !== password_database) {
        console.log(username, password);
        return res.status(401).send('Invalid credentials');
    }

    const token = generateToken(username);
    res.cookie('token', token, { maxAge: jwtExpirySeconds * 1000 });
    res.end();
});

app.get('/protected', (req, res) => {
    const token = req.cookies.token;

    if (!token) {
        return res.status(401).send('No token');
    }

    const payload = validateToken(token);

    if (!payload) {
        return res.status(401).send('Invalid token');
    }

    res.send(`Hello, ${payload.username}!`);
});

app.listen(8080, () => {
    console.log('JWT Server One Route started at :8080');
});
